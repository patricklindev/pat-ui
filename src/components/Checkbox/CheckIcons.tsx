export interface IIconPath {
  [name: string]: {
    path: string;
    viewBox: string;
  };
}

export const CheckIcons: IIconPath = {
  ['checkbox-outline']: {
    path: 'M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z',
    viewBox: '0 0 24 24',
  },
  ['checkbox-fill']: {
    path: 'M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z',
    viewBox: '0 0 24 24',
  },
  ['heart-outline']: {
    path: 'M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z',
    viewBox: '0 0 24 24',
  },
  ['heart-fill']: {
    path: 'm12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z',
    viewBox: '0 0 24 24',
  },
  ['bookmark-outline']: {
    path: 'M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15-5-2.18L7 18V5h10v13z',
    viewBox: '0 0 24 24',
  },
  ['bookmark-fill']: {
    path: 'M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z',
    viewBox: '0 0 24 24',
  },
};
