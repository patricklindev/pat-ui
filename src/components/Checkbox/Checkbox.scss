// checkbox wrapper style 
.basicCheckboxWrapper,
.diffIconCheckboxWrapper,
.diffThemeCheckboxWrapper,
.diffIconThemeCheckboxWrapper,
.checkboxWithLabelWrapper,
.checkboxDisableWrapper{
@include displayWrapper
}

// hide default checkbox
.checkbox-container input[type="checkbox"]{
  cursor: pointer;
  opacity:0;
  position: absolute;
}

[data-testid='iconSpan']{
  position: relative;
}

.checkbox-container{
  display:flex;
  align-items:center;
  margin-right: 1rem;
  margin: 1rem;
}

.checkbox-container label{
  cursor:pointer;
  display:flex;
  justify-content: center;
  align-items: center;
  position: relative;
}

// .checkbox-container label span::before{
  
//   content: " "; /* :after will only appear if this content is declared */
//   width: 3em;
//   height: 3em;
//   border-radius: 50%;
//   background-color: rgba($color: tomato,$alpha: .1);
//   position: absolute;
//   opacity: 0;
//   transition: opacity .1s ease-in;
// }

.checkbox-container label:hover span::before,
.checkbox-container input[type='checkbox']:hover + label span::before{
  visibility:visible;
  opacity: 1;
}


.checkbox-container label:active span::before,
.checkbox-container input[type='checkbox']:active + label span::before{
  opacity: 1;
  visibility:visible;
}

.checkbox-container label span{
  border-radius:.18em;
  border: .15em solid gray;
}

.checkbox{
  display: flex;
  justify-content: center;
  align-items: center;
  &::before{
    @include checkboxHover($gray-500)
  } 

  &-ex-small{
    @include check-box-size($cb-ex-small)
  }

  &-small{
    @include check-box-size($cb-small)
  }

  &-normal{
    @include check-box-size($cb-normal)
  }

  &-large{
    @include check-box-size($cb-large)
  }

  &-ex-larger{
    @include check-box-size($cb-ex-large)
  }

  &-primary{
    background-color: $primary; 
  }
}


.bg-iChecked-path{
  &-primary{
    fill: $white; 
  }

  &-secondary{
    fill: $white; ; 
  }

  &-success{
    fill: $white; ; 
  }

  &-info{
    fill: $white ; 
  }

  &-warning{
    fill: $white; ; 
  }

  &-danger{
    fill: $white; 
  }

  &-default{
    fill: invert($white); 
  }

  &-dark{
    fill: $white; ; 
  }
}

.bg-iCheck-path{
  &-primary{
    fill: $primary; 
  }

  &-secondary{
    fill: $secondary; 
  }

  &-success{
    fill: $success; 
  }

  &-info{
    fill: $info;
  }

  &-warning{
    fill:$warning;
  }

  &-danger{
    fill: $danger;
  }

  &-default{
    fill: $white;
  }

  &-dark{
    fill: $dark;
  }

  &-disable{
    color: gray;
    cursor:not-allowed;
    color: #000;
  }
}

.bg-iOther-path-default{
  fill: $gray-500;
}

.bg-iOther-check-path{
  &-primary{
    fill: $primary; 
  }

  &-secondary{
    fill: $secondary; 
  }

  &-success{
    fill: $success; 
  }

  &-info{
    fill: $info;
  }

  &-warning{
    fill:$warning;
  }

  &-danger{
    fill: $danger;
  }

  &-default{
    fill: $black;
  }

  &-dark{
    fill: $dark;
  }

  &-disable{
    color: gray;
    cursor:not-allowed;
    color: #000;
  }
}


.bg-iOther-default{
  border: none !important; 
  border-radius: 30% !important;
}


.bg-iCheck{
  &-primary{
    background-color: $primary; 
    &::before{
      @include checkboxHover($primary)
    }
  }

  &-secondary{
    background-color: $secondary; 
    &::before{
      @include checkboxHover($secondary)
    }
  }

  &-success{
    background-color: $success;
    &::before{
      @include checkboxHover($success)
    } 
  }

  &-info{
    background-color: $info;
    &::before{
      @include checkboxHover($info)
    } 
  }

  &-warning{
    background-color:$warning;
    &::before{
      @include checkboxHover($warning)
    } 
  }

  &-danger{
    background-color: $danger;
    &::before{
      @include checkboxHover($danger)
    } 
  }

  &-default{
    background-color: $white;
    &::before{
      @include checkboxHover($gray-500)
    } 
  }

  &-dark{
    background-color: $dark;
    &::before{
      @include checkboxHover($dark)
    } 
  }

  &-disable{
    color: gray;
    cursor:not-allowed;

    color: #000;
  }
}

.bg-span-iOther-hover{
  &-primary{ 
    &::before{
      @include checkboxHover($primary)
    }
  }

  &-secondary{
    &::before{
      @include checkboxHover($secondary)
    }
  }

  &-success{
    &::before{
      @include checkboxHover($success)
    } 
  }

  &-info{
    &::before{
      @include checkboxHover($info)
    } 
  }

  &-warning{
    &::before{
      @include checkboxHover($warning)
    } 
  }

  &-danger{
    &::before{
      @include checkboxHover($danger)
    } 
  }

  &-default{
    &::before{
      @include checkboxHover($gray-500)
    } 
  }

  &-dark{
    &::before{
      @include checkboxHover($dark)
    } 
  }
}