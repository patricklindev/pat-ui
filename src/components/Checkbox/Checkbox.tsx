import React,{useState,FC,useEffect,useRef}from 'react'
import { classNames } from '../../utils/classNames';
import { uid } from '../../utils/uuid';
import { IconPath } from './Icons';

export type boxSize = 'ex-small'|'small' | 'normal' | 'large' | 'ex-larger';
export type iconType = 'home' | 'spinner' | 'angle down' | 'plus' | 'home' | 'users' | 'times' | 'search' | 'star' | 'moon' | 'heart' | 'smile wink' | 'truck' | 'credit card' | 'check' | 'batch';
export type themeColor = 'primary' | 'secondary' | 'success' | 'info' | 'warning' | 'danger' | 'dark' | 'default';

export interface ICheckboxProps {
  /** set customized style */
  className?: string;
  /** set checkbox size */
  checkSize?: boxSize
  /** make checkbox bar always check */
  checked?: boolean ;
  /** add specific icon to input */
  icon?: iconType;
  /** add specific icon theme to input */
  iconTheme?: themeColor;
  /**customize checkbox color */
  checkBgTheme? : themeColor;
  //**set background theme when check on check icon */
  checkedBgTheme? : themeColor;
  /** pass a callback function out-site of props */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /** set label value */ 
  label?: string;
}

/**
 * Checkbox allow user to implement different styles and features
 *
 * ```js
 * import { Checkbox } from 'pat-ui'
 * ```
 */

export const Checkbox:FC<ICheckboxProps> = (props)=> {

  const { checkSize, checked, checkBgTheme,checkedBgTheme, icon, iconTheme,label,onChange } = props;

  // toggle the checked state
  const [isCheck,setIsCheck] = useState(checked ? checked : false)

  // define a unique id when the function is mounted
  const [id,setId] = useState(uid())

  let classNameSpan;

  //if user defined a background theme on component for checked state
  if(checkedBgTheme){
    classNameSpan = {
      //class for control box size if user define checkSize prop
      [`checkbox-${checkSize}`]: !!checkSize,
      //change the background color depend on checkedBgTheme when user checked
      [`bg-iCheck-${checkedBgTheme}`]: isCheck,
    }
  }else{ //when user not provide any background theme for check state

    if(icon !== 'check'){ //if the icon is not check icon
      classNameSpan = {
        //class for control box size
        [`checkbox-${checkSize}`]: !!checkSize,
        // remove the span border
        'bg-iOther-default': true,
      }
    }else{ // if the icon is check icon
      classNameSpan = {
        // class for control box size
        [`checkbox-${checkSize}`]: !!checkSize,
        // class that is control the background color when user pass checkBgTheme prop for check icon when uncheck
        [`bg-iCheck-${checkBgTheme}`]: !!checkBgTheme,
      }
    }
  }


  let classNameSVG;

  // when user define a background theme on check
  if(checkedBgTheme){
    classNameSVG = {
      // define the check icon path value depends on the checkBgTheme when uncheck 
      [`bg-iCheck-path-${checkBgTheme}`]: !isCheck,

      // define the check icon path value depends on the checkBgTheme when check 
      [`bg-iChecked-path-${checkedBgTheme}`]: isCheck,
    }
  }else{ // when user don't provide checkedBgTheme
    if(icon !== 'check'){ // if the icon is not check
      if(iconTheme){
        classNameSVG = {
          //set default icon color when uncheck
          'bg-iOther-path-default': !isCheck,
          // set icon color when check
          [`bg-iOther-check-path-${iconTheme}`]: isCheck
        }
        // add hover style with before pseudo element when use provide iconTheme
        classNameSpan = {
          ...classNameSpan,
          [`bg-span-iOther-hover-${iconTheme}`]: isCheck
        }
      }else{ // if you don't define icon color
        classNameSVG = {
          // set both check and uncheck to be default color
          'bg-iOther-path-default': !isCheck,
          'bg-iOther-check-path-default': isCheck
        }
      }
    }else{ // if the icon is check
      classNameSVG = {
        //set check icon uncheck color depends on checkBgTheme props
        [`bg-iCheck-path-${checkBgTheme}`]: !isCheck,
        //set check icon check color depends on checkBgTheme props
        [`bg-iChecked-path-${checkBgTheme}`]: isCheck,
      }
    }
  }
  
    // add class name for SVG
  const styleClassNameSVG = classNames(classNameSVG)
    // add class name for Span
  const styleClassNameSpan = classNames('checkbox',classNameSpan)

  const handleChange = (e:React.ChangeEvent<HTMLInputElement>) => {
      setIsCheck(e.target.checked)
  }

  useEffect(()=>{
    if(checked !== undefined){
      setIsCheck(checked)
    }
  },[checked])

  return (
    <div className='checkbox-container'>
      <input type='checkbox' onChange={onChange ? onChange : handleChange} checked={isCheck} id={id} data-testid='inputCheckBox'/> 
      <label htmlFor={id} data-testid='iconLabel' className='labelContent'>
        <span className={styleClassNameSpan} data-testid='iconSpan'>
          <svg viewBox={IconPath[`${icon}`].viewBox} role="img" data-testid='iconSVG'>
            <path d={IconPath[`${icon}`].path} className={styleClassNameSVG} data-testid='iconSVGPath'/>
          </svg>
        </span>
        {label}
      </label> 
    </div>
  )
}

Checkbox.defaultProps = {
  checkBgTheme : 'default',
  checkSize: 'normal',
  icon: 'check'
}

export default Checkbox;