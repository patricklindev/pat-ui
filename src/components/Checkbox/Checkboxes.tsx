/* eslint-disable no-useless-computed-key */
export interface CheckboxPath {
  [name: string]: {
    checkedPath: string;
    uncheckedPath: string;
    indeterminatePath?: string;
    viewBox: string;
  };
};

export const CheckboxPaths: CheckboxPath = {
  ['checkbox']: {
    checkedPath: 'M 19 3 H 5 c -1.11 0 -2 0.9 -2 2 v 14 c 0 1.1 0.89 2 2 2 h 14 c 1.11 0 2 -0.9 2 -2 V 5 c 0 -1.1 -0.89 -2 -2 -2 Z m -9 14 l -5 -5 l 1.41 -1.41 L 10 14.17 l 7.59 -7.59 L 19 8 l -9 9 Z',
    uncheckedPath: 'M 19 5 v 14 H 5 V 5 h 14 m 0 -2 H 5 c -1.1 0 -2 0.9 -2 2 v 14 c 0 1.1 0.9 2 2 2 h 14 c 1.1 0 2 -0.9 2 -2 V 5 c 0 -1.1 -0.9 -2 -2 -2 Z',
    indeterminatePath: 'M 19 3 H 5 c -1.1 0 -2 0.9 -2 2 v 14 c 0 1.1 0.9 2 2 2 h 14 c 1.1 0 2 -0.9 2 -2 V 5 c 0 -1.1 -0.9 -2 -2 -2 Z m -2 10 H 7 v -2 h 10 v 2 Z',
    viewBox: '0 0 24 24',
  },
  ['favorite']: {
    checkedPath: 'm 12 21.35 l -1.45 -1.32 C 5.4 15.36 2 12.28 2 8.5 C 2 5.42 4.42 3 7.5 3 c 1.74 0 3.41 0.81 4.5 2.09 C 13.09 3.81 14.76 3 16.5 3 C 19.58 3 22 5.42 22 8.5 c 0 3.78 -3.4 6.86 -8.55 11.54 L 12 21.35 Z',
    uncheckedPath: 'M 16.5 3 c -1.74 0 -3.41 0.81 -4.5 2.09 C 10.91 3.81 9.24 3 7.5 3 C 4.42 3 2 5.42 2 8.5 c 0 3.78 3.4 6.86 8.55 11.54 L 12 21.35 l 1.45 -1.32 C 18.6 15.36 22 12.28 22 8.5 C 22 5.42 19.58 3 16.5 3 Z m -4.4 15.55 l -0.1 0.1 l -0.1 -0.1 C 7.14 14.24 4 11.39 4 8.5 C 4 6.5 5.5 5 7.5 5 c 1.54 0 3.04 0.99 3.57 2.36 h 1.87 C 13.46 5.99 14.96 5 16.5 5 c 2 0 3.5 1.5 3.5 3.5 c 0 2.89 -3.14 5.74 -7.9 10.05 Z',
    indeterminatePath: 'M 16.5 3 c -1.74 0 -3.41 0.81 -4.5 2.09 C 10.91 3.81 9.24 3 7.5 3 C 4.42 3 2 5.42 2 8.5 c 0 3.78 3.4 6.86 8.55 11.54 L 12 21.35 l 1.45 -1.32 C 18.6 15.36 22 12.28 22 8.5 C 22 5.42 19.58 3 16.5 3 Z m -4.4 15.55 l -0.1 0.1 l -0.1 -0.1 C 7.14 14.24 4 11.39 4 8.5 C 4 6.5 5.5 5 7.5 5 c 1.54 0 3.04 0.99 3.57 2.36 h 1.87 C 13.46 5.99 14.96 5 16.5 5 c 2 0 3.5 1.5 3.5 3.5 c 0 2.89 -3.14 5.74 -7.9 10.05 Z',
    viewBox: '0 0 24 24',
  },
  ['bookmark']: {
    checkedPath: 'M 17 3 H 7 c -1.1 0 -1.99 0.9 -1.99 2 L 5 21 l 7 -3 l 7 3 V 5 c 0 -1.1 -0.9 -2 -2 -2 Z',
    uncheckedPath: 'M 17 3 H 7 c -1.1 0 -1.99 0.9 -1.99 2 L 5 21 l 7 -3 l 7 3 V 5 c 0 -1.1 -0.9 -2 -2 -2 Z m 0 15 l -5 -2.18 L 7 18 V 5 h 10 v 13 Z',
    indeterminatePath: 'M 17 3 H 7 c -1.1 0 -1.99 0.9 -1.99 2 L 5 21 l 7 -3 l 7 3 V 5 c 0 -1.1 -0.9 -2 -2 -2 Z m 0 15 l -5 -2.18 L 7 18 V 5 h 10 v 13 Z',
    viewBox: '0 0 24 24',
  },
};