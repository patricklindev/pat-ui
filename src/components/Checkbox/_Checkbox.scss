.cb {
  position: relative;
  display: inline-block;
  font-weight: $cb-font-weight;
  line-height: $cb-line-height;
  color: $body-color;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  cursor: pointer;
  transition: $cb-transition;

  // cb size
  @include cb-size(
    $cb-font-size,
    $cb-box-margin-right,
    $cb-box-transform-scale
  );
}

.cb-sm {
  @include cb-size(
    $cb-font-size-sm,
    $cb-box-margin-right-sm,
    $cb-box-transform-scale-sm
  );
}
.cb-lg {
  @include cb-size(
    $cb-font-size-lg,
    $cb-box-margin-right-lg,
    $cb-box-transform-scale-lg
  );
}

.cb-primary {
  @include cb-style($white, $primary, $primary);
}
.cb-secondary {
  @include cb-style($white, $secondary, $secondary);
}
.cb-danger {
  @include cb-style($white, $danger, $danger);
}
.cb-info {
  @include cb-style($white, $info, $info);
}
.cb-success {
  @include cb-style($white, $success, $success);
}
.cb-warning {
  @include cb-style($white, $warning, $warning);
}

.cb-icon {
  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
  }
}
