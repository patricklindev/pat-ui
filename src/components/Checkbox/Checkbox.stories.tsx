import React,{useState}  from 'react';
import { action } from '@storybook/addon-actions';
import Checkbox from './Checkbox';

export default {
  title: 'Checkbox',
  component: Checkbox,
};

export const BasicCheckboxes = () => {
  return (
    <div className='basicCheckboxWrapper'>
      <Checkbox />
      <Checkbox checkBgTheme='secondary'/>
      <Checkbox checkBgTheme='primary'/>
      <Checkbox checkSize='large'/>
    </div>
  );
}

export const DiffThemeCheckbox = () => (
  <div className='diffThemeCheckboxWrapper'>
    <Checkbox checkedBgTheme="primary" checked />
    <Checkbox checkedBgTheme="secondary" checked/>
    <Checkbox checkedBgTheme="success" checked/>
    <Checkbox checkedBgTheme="info" checked/>
    <Checkbox checkedBgTheme="warning" checked/>
    <Checkbox checkedBgTheme="danger" checked/>
    <Checkbox checkedBgTheme="default" checked/>
    <Checkbox checkedBgTheme="dark" checked/>
  </div>
);

export const DiffIconCheckbox = () => (
  <div className='diffIconCheckboxWrapper'>
    <Checkbox icon="home" />
    <Checkbox icon="spinner" />
    <Checkbox icon="angle down" />
    <Checkbox icon="plus" />
    <Checkbox icon="home" />
    <Checkbox icon="users" />
    <Checkbox icon="times" />
    <Checkbox icon="search" />
    <Checkbox icon="star" />
    <Checkbox icon="moon" />
    <Checkbox icon="heart" />
    <Checkbox icon="smile wink" />
    <Checkbox icon="truck" />
    <Checkbox icon="credit card" />
    <Checkbox icon="batch" />
  </div>
);

export const DiffIconThemeCheckbox = () => (
  <div className='diffIconThemeCheckboxWrapper'>
    <Checkbox icon="home" iconTheme="primary" checked/>
    <Checkbox icon="spinner" iconTheme="secondary" checked/>
    <Checkbox icon="angle down" iconTheme="success" checked/>
    <Checkbox icon="plus" iconTheme="info" checked/>
    <Checkbox icon="home" iconTheme="warning" checked/>
    <Checkbox icon="users" iconTheme="danger" checked/>
    <Checkbox icon="search" iconTheme="dark" checked/>
    <Checkbox icon="batch" iconTheme="primary" checked/>
  </div>
);

export const CheckboxWithLabel = () => (
  <div className='checkboxWithLabelWrapper'>
    <Checkbox label='Hello,World' />
    <Checkbox icon="home" checked label='Hello,World'/>
    <Checkbox icon="users" iconTheme="warning" checked label='Hello,World'/>
  </div>
);


export const ControlCheckStateByCallbackFunc = () => {
  const [isCheck,setIsCheck] = useState(false)

  const handleChange = (event:React.ChangeEvent<HTMLInputElement>)=>{
    setIsCheck(event.target.checked)
    console.log('GG from outSite')
  }

  return (
    <div className='basicCheckboxWrapper'>
      <Checkbox icon='home' iconTheme='primary' checked={isCheck} onChange={handleChange} />
    </div>
  );
}