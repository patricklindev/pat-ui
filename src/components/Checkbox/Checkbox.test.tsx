import React from 'react';
import { render,fireEvent } from '@testing-library/react';
import Checkbox from './Checkbox';
import { IconPath } from './Icons';
import {ICheckboxProps} from './Checkbox';


it('should match snapshot',()=>{
  const {asFragment} = render(<Checkbox />)
  expect(asFragment()).toMatchSnapshot()
})

describe('<Checkbox />',()=>{

  it('should able to change the check/uncheck state by default',()=>{
    const {getByTestId}  = render(<Checkbox />)

    const checkBoxInput = getByTestId('inputCheckBox')
    const iconSVG = getByTestId('iconSVG')

    expect(checkBoxInput).toBeInTheDocument()
    expect(iconSVG).toBeInTheDocument()
    expect(iconSVG.firstChild).not.toHaveClass('bg-iChecked-path-default')

    fireEvent.click(checkBoxInput)
    expect(iconSVG.firstChild).toHaveClass('bg-iChecked-path-default')
  })

  it('should be able to choose the size of the component with some predefined size from props',()=>{
    const {getByTestId}  = render(<Checkbox checkSize='large'/>)
    const targetElement = getByTestId('iconSpan')
    expect(targetElement).toBeInTheDocument()
    expect(targetElement).toHaveClass('checkbox-large')
  })

  test('when user wont defined any checkSize value, checkSize value should be normal',()=>{
    const {getByTestId}  = render(<Checkbox />)
    const targetElement = getByTestId('iconSpan')
    expect(targetElement).toBeInTheDocument()
    expect(targetElement).toHaveClass('checkbox-normal')
  })

  it('should be able to choose the color of the component with some predefine color from props when uncheck',()=>{
    const {getByTestId}  = render(<Checkbox checkBgTheme='secondary'/>)
    const targetElement = getByTestId('iconSpan')
    expect(targetElement).toBeInTheDocument()
    expect(targetElement).toHaveClass('bg-iCheck-secondary')
  })


  test('svg fill value will be the checkBgTheme theme provide from props when uncheck',()=>{
    const {getByTestId}  = render(<Checkbox checkBgTheme='secondary'/>)

    const iconSVG = getByTestId('iconSVG')
    expect(iconSVG.firstChild).toHaveClass('bg-iCheck-path-secondary')
  })

  test('svg fill value will be the checkBgTheme theme provide from props when check',()=>{
    const {getByTestId}  = render(<Checkbox checkBgTheme='secondary'/>)

    const checkBoxInput = getByTestId('inputCheckBox')
    const iconSVG = getByTestId('iconSVG')
    expect(iconSVG.firstChild).not.toHaveClass('bg-iChecked-path-secondary')

    fireEvent.click(checkBoxInput)
    expect(iconSVG.firstChild).toHaveClass('bg-iChecked-path-secondary')
  })

  it('should be the default checkBgTheme when user don\'t provide checkBgTheme props',()=>{
    const {getByTestId}  = render(<Checkbox />)

    const targetElement = getByTestId('iconSpan')
    expect(targetElement).toHaveClass('bg-iCheck-default')
  })

  it('should be the default svg fill value when user don\'t provide checkBgTheme props when uncheck',()=>{
    const {getByTestId}  = render(<Checkbox />)

    const iconSVG = getByTestId('iconSVG')
    expect(iconSVG.firstChild).toHaveClass('bg-iCheck-path-default')
  })

  it('should be the default svg fill value when user don\'t provide checkBgTheme props when check',()=>{
    const {getByTestId}  = render(<Checkbox />)

    const checkBoxInput = getByTestId('inputCheckBox')
    const iconSVG = getByTestId('iconSVG')
    expect(iconSVG.firstChild).not.toHaveClass('bg-iChecked-path-default')
    fireEvent.click(checkBoxInput)
    expect(iconSVG.firstChild).toHaveClass('bg-iChecked-path-default')
  })

  it('should be able to choose the color of component with some predefine color from props when check',()=>{
    const {getByTestId} = render(<Checkbox checkedBgTheme='secondary'/>)

    const checkBoxInput = getByTestId('inputCheckBox')
    const targetElement = getByTestId('iconSpan')
    const iconSVG = getByTestId('iconSVG')

    expect(targetElement).not.toHaveClass('bg-iCheck-secondary')
    expect(iconSVG.firstChild).not.toHaveClass('bg-iChecked-path-secondary')

    fireEvent.click(checkBoxInput)

    expect(targetElement).toHaveClass('bg-iCheck-secondary')
    expect(iconSVG.firstChild).toHaveClass('bg-iChecked-path-secondary')
  })

  it('should be able to decide the icons to represent check/unchecked state from props',()=>{
    const {getByTestId} = render(<Checkbox icon='home'/>)

    const iconSVG = getByTestId('iconSVG')
    const iconSVGPath = getByTestId('iconSVGPath')

    expect(iconSVG.getAttribute('viewBox')).toBe(IconPath['home'].viewBox)
    expect(iconSVGPath.getAttribute('d')).toBe(IconPath['home'].path)
  })

  test('when user define icon props, not define iconTheme props, svg path fill value will be default when uncheck',()=>{
    const {getByTestId} = render(<Checkbox icon='home'/>)

    const iconSVG = getByTestId('iconSVG')
    expect(iconSVG.firstChild).toHaveClass('bg-iOther-path-default')
  })

  test('when user define icon props, not define iconTheme props, svg path fill value will be default when check',()=>{
    const {getByTestId} = render(<Checkbox icon='home'/>)

    const iconSVG = getByTestId('iconSVG')
    const checkBoxInput = getByTestId('inputCheckBox')

    expect(iconSVG.firstChild).not.toHaveClass('bg-iOther-check-path-default')

    fireEvent.click(checkBoxInput)

    expect(iconSVG.firstChild).toHaveClass('bg-iOther-check-path-default')
  })

  test('when user define iconTheme and icon props,svg path fill value should be iconTheme value when check',()=>{
    const {getByTestId} = render(<Checkbox iconTheme = 'primary' icon='home'/>)

    const checkBoxInput = getByTestId('inputCheckBox')
    const iconSVG = getByTestId('iconSVG')

    expect(iconSVG.firstChild).not.toHaveClass('bg-iOther-check-path-primary')
    fireEvent.click(checkBoxInput)
    expect(iconSVG.firstChild).toHaveClass('bg-iOther-check-path-primary')
  })

  test('when user define icon props, svg wrapper will remove border',()=>{
    const {getByTestId} = render(<Checkbox icon='home'/>)

    const targetElement = getByTestId('iconSpan')
    expect(targetElement).toHaveClass('bg-iOther-default')
  })

  it('should be able to define the label from the props',()=>{
    const {getByText} = render(<Checkbox label='hello,world'/>)

    expect(getByText('hello,world')).toBeInTheDocument()
  })

  it('can control the check/uncheck state from the props',()=>{
    const {getByTestId} = render(<Checkbox checked/>)
    const checkBoxInput = getByTestId('inputCheckBox')

    expect(checkBoxInput).toBeChecked()
  })

  it('can listen to the change of check/uncheck state from outside of the component by passing the onChange callback from props',()=>{

    const btnLinkProps: ICheckboxProps = {
      onChange: jest.fn(),
    };

    const {getByTestId} = render(<Checkbox {...btnLinkProps}/>)

    const checkBoxInput = getByTestId('inputCheckBox')

    fireEvent.change(checkBoxInput, {target: {checked: false}});
    expect(checkBoxInput).not.toBeChecked()

    fireEvent.change(checkBoxInput, {target: {checked: true}});
    expect(checkBoxInput).toBeChecked()
  })
})
