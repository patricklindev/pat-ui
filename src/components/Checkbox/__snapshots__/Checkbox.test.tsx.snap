// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <input
      data-testid="inputCheckBox"
      id="l5jv57zp4xxls9rf7vn"
      type="checkbox"
    />
    <label
      class="labelContent"
      data-testid="iconLabel"
      for="l5jv57zp4xxls9rf7vn"
    >
      <span
        class="checkbox checkbox-normal bg-iCheck-default"
        data-testid="iconSpan"
      >
        <svg
          data-testid="iconSVG"
          role="img"
          viewBox="0 0 512 512"
        >
          <path
            class="bg-iCheck-path-default"
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            data-testid="iconSVGPath"
          />
        </svg>
      </span>
    </label>
  </div>
</DocumentFragment>
`;
